src/workerd/api/worker-loader.h

```diff
@@ -0,0 +1,128 @@
+// specify arbitrary props).
+class WorkerStub: public jsg::Object {
+ public:
+  WorkerStub(IoOwn<WorkerStubChannel> channel): channel(kj::mv(channel)) {}
+
+  struct EntrypointOptions {
+    jsg::Optional<jsg::JsObject> props;
+
+    JSG_STRUCT(props);
+  };
+
+  jsg::Ref<Fetcher> getEntrypoint(jsg::Lock& js,
+      jsg::Optional<kj::Maybe<kj::String>> name,
+      jsg::Optional<EntrypointOptions> options);
+  jsg::Ref<DurableObjectClass> getDurableObjectClass(jsg::Lock& js,
+      jsg::Optional<kj::Maybe<kj::String>> name,
+      jsg::Optional<EntrypointOptions> options);
+
+  JSG_RESOURCE_TYPE(WorkerStub) {
+    JSG_METHOD(getEntrypoint);
+    JSG_METHOD(getDurableObjectClass);
+  }
+
+ private:
+  IoOwn<WorkerStubChannel> channel;
+};
+
+// JS interface for worker loader binding.
+class WorkerLoader: public jsg::Object {
+ public:
+  // Create a WorkerLoader backed by the given I/O channel.
+  //
+  // `compatDateValidation` will differ between workerd vs. production.
+  explicit WorkerLoader(uint channel, CompatibilityDateValidation compatDateValidation)
+      : channel(channel),
+        compatDateValidation(compatDateValidation) {}
+
+  struct Module {
+    // Exactly one must be filled in.
+    jsg::Optional<kj::String> js;               // ES module
+    jsg::Optional<kj::String> cjs;              // Common JS module
+    jsg::Optional<kj::String> text;             // text blob, imports as a string
+    jsg::Optional<kj::Array<const byte>> data;  // byte blob, imports as ArrayBuffer
+    jsg::Optional<jsg::Value> json;             // arbitrary JS value, will be serialized to JSON
+                                                // and then parsed again when imported
+
+    JSG_STRUCT(js, cjs, text, data, json);
+
+    // HACK: When we serialize the JSON in extractSource() we need to place the owned kj::String
+    //   somewhere since Worker::Script::Source only gets a kj::StringPtr.
+    kj::Maybe<kj::String> serializedJson;
+  };
+
+  struct WorkerCode {
+    kj::String compatibilityDate;
+    jsg::Optional<kj::Array<kj::String>> compatibilityFlags;
+    jsg::Optional<bool> allowExperimental = false;
+
+    kj::String mainModule;
+
+    // Modules are specified as an object mapping names to content. If the content is just a
+    // string, an ES module is assumed. If it's an object, the type of module is determined
+    // based on which property is set.
+    jsg::Dict<kj::OneOf<Module, kj::String>> modules;
+
+    // Any RPC-serializable value!
+    jsg::Optional<jsg::JsRef<jsg::JsObject>> env;
+
+    // `Fetcher` (e.g. service binding) representing the loaded worker's global outbound.
+    //
+    // If omitted, inherit the current worker's global outbound.
+    //
+    // If `null`, block the global outbound (all requests throw errors).
+    jsg::Optional<kj::Maybe<jsg::Ref<Fetcher>>> globalOutbound;
+
+    // TODO(someday): cache API outbound?
+
+    // TODO(someday): Tails?
```

Comment:
I'm assuming this means tail workers for observability? If so, it would be worth expanding the comment just a bit to make it clearer